### DISCUSSION TITLE
How to specify the constraints file for a test?

### DISCUSSION THREAD

**Discussion by brendan-morin** [tox-dev/tox#3074] [Q&A] [answered]
Let's say I have a package that has `install_requires=[pyspark >= 3.0]` (among other requirements). I want to run tests on this package, but enforce a constraint on the install of it into the test environment. Let's say I want to constrain to `pyspark=3.2.0`, but allow the other requirements to resolve as they would naturally.

I tried passing in a constraints file with this constraint to `deps` and setting `constrain_package_deps=true`, but I get the error `ERROR: Constraints cannot have extras`. I checked `{env}/constraints.txt`, and I see it listed all of my unpinned requirements from install_requires in there (including one with extras, hence the error). That feels like not at all what I am looking for. 

Is there a way to specify exactly what ends up in constraints.txt?

**Comment by brendan-morin** [comment]
Weirdly, it seems like the best way I've found to do this is to actually pre-install my constraints file as part of deps, rather than using it as a constraint:
```
deps =  -r requirements/test-constraints.txt
```
If I use it as an actual constraint (`-c`) in `deps`, it's overridden by the subsequent package install.

I don't want to use all of the dependencies in `deps` as my constraints (which seems to be my other option), since those are my test dependencies. If anything, those should be constrained based on my package dependencies. Ideally, the constraint chain should follow:

```
user defined constraints file -> constrains package deps -> constrains test deps
```

Am I misunderstanding how this works? It seems like the above workaround is the only way to get what I'm looking for.

**Reply by gaborbernat** [reply]
Did you read https://tox.wiki/en/4.6.4/faq.html#using-constraint-files? tox itself does not offer constraint support, besides passing flags to pip. So ultimately you need to convince pip to do what you want, not tox per se.

**Reply by brendan-morin** [reply]
Yes, I read this, and from what I can tell, the described use case is not addressed. The problem is that not every install command tox does seems to allow passing flags to pip. If I were to replicate what I would intend to be done, in pure pip commands:

```bash
# install the library and all requirements, as a user might + constraints
pip install . -c my_constraint.txt

# freeze the venv to use this as a constraint for test requirements
# we want to do this because the test requirements should not be able to
# modify or override the install requirements
pip freeze > test_constraints.txt

# we now have a constraint file that represents a real world constrained 
# install - now we install the test requirements around that (pytest, linters, etc)
pip install -r test_requirements.txt -c test_constraints.txt
```

**Reply by brendan-morin** [reply]
I don't expect to pip freeze in the middle of tox's workflow, but I've added that in to demonstrate the intent is to constrain the library install (as a user might) and then test that state, while avoiding test requirements from mutating this state.

**Reply by brendan-morin** [reply]
I'm currently exploring if the [install_command](https://tox.wiki/en/4.6.4/config.html#pip-installer) is appropriate... some more documentation on this might be helpful. Currently trying:
```
install_command = python -I -m pip install <opts> <packages> -c my_constraints.txt
```
But that doesn't seem to substitute the packages into `<packages>`, but rather insert it after. I've also tried `{packages}`, both fail like:
```
'python -I -m pip install {opts} {packages} -c my_constraints.txt' coverage pytest ... (etc)
```

**Reply by brendan-morin** [reply, marked as answer]
`install_command` is the answer. Using `constrain_package_deps` or `use_frozen_constraints` was barking up the wrong tree. It wasn't explicit in the documentation, but this is applied to both `dep` and package installs. If using this in `tox.ini`, this will work:
```
install_command =
    python -I -m pip install {opts} {packages} -c my_constraints.txt
```

The only thing that would be one better would be being able to have separate install commands for deps and package installs (so they can get their own constraints). Perhaps a feature request.

**Reply by gaborbernat** [reply]
`install_command` is deprecated, you should use setting environment variables instead to control pip via `PIP_` 😊 one of the reasons we removed https://tox.wiki/en/4.6.4/upgrading.html#cli-arguments-changed the `--index-url` flag.

**Reply by brendan-morin** [reply]
I read through the docs there, but I'm not sure I understand what actionable next steps are. Any chance you could help me understand how I could use `PIP_` for the above use case instead of `install_command`? Also maybe I'm just bad at searching, but I don't see where theres mention in the docs about `install_command` being deprecated.

**Reply by gaborbernat** [reply]
This should be equivalent:

```ini
set_env = 
	PIP_CONSTRAINT=my_constraints.txt
```

> Also maybe I'm just bad at searching, but I don't see where theres mention in the docs about `install_command` being deprecated.

Yeah, TBD, but I do plan to drop it.

**Reply by brendan-morin** [reply]
Great! That does appear to work, thank you. One (minor) downside I'm noting vs `install_command` is that this constraints file is no longer mentioned in the logs. Perhaps an extra log setting to output `PIP_` vars used could be a helpful (though there may be some security concerns, e.g. username/passwords in index URLs), or something similar just to confirm that what we think is happening is actually happening.

**Reply by brendan-morin** [reply]
Apologies for continuing to poke on this issue, but I want to make sure you're aware of some of the nuance here as you think about how you want to evolve tox and handle (or not) these various cases.

While switching to `PIP_CONSTRAINT`, things initially felt fine for the one pipeline I ran in isolation. However, I quickly ran into issues, as the scope of `PIP_CONSTRAINT` is broader than just the install commands. For example, I have a tox test/env that runs my fix/lint commands as pre-commit hooks. When pre-commit installs its hooks, it runs a pip install command for each hook (at least for python), as each hook gets its own mini env. `PIP_CONSTRAINT` was leaking into these pre-commit setup commands, causing unintended consequences (in my case it wasn't even able to find the referenced file, and auto-failed). 

This was compounded by there not being a great way to unset environment variables (https://github.com/tox-dev/tox/issues/1387). I fiddled unsuccessfully with this a bit, and there probably is some way forward here, but ultimately reverting to `install_comand` was the quickest and most effective way to isolate the scope to just tox's installs. I guess I'll plan to deal with it working around this deprecating at some point in the future whenever it happens.

I understand why this can be seen as a feature, but I think it's reasonable to want to just control where constraints get applied for the exact commands they are applied to. In my case this led to a fair bit of confusion, and ultimately didn't feel like the most straightforward to continue to make use of the recommendation to set `PIP_` env vars, due to the potential for unintended side effects.

**Reply by gaborbernat** [reply]
I think being able to adjust the scope of the environment variables feels like a good idea. PR welcome to add this functionality.

**Reply by webknjaz** [reply]
@BrendanJM you hit something that many people don't understand — the more envs you have the harder it is to have a combined constraints (lock) file.

Check out my experiments/PoCs/thoughts with having per-env lockfiles @ https://github.com/jazzband/pip-tools/issues/826#issuecomment-1324127670 & https://github.com/jazzband/pip-tools/issues/1326#issuecomment-1834517252. The first link has examples of wrappers for `pip install` that auto-injects constraints, for example. This is mostly tox 3-oriented, though. In some places I was resetting the env vars via something like `setenv = PIP_CONSTRAINTS=`, by the way. And a lot of times you can get away with `-c` in `deps =` if you know what you're doing :)
